internal/keys/keys.go

```diff
@@ -0,0 +1,158 @@
+package keys
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	cryptorand "crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"math"
+	"math/big"
+	"time"
+
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+)
+
+const duration365d = time.Hour * 24 * 365
+
+// SignerWithPEM is a signer implementation that has its PEM encoded representation attached
+type SignerWithPEM interface {
+	crypto.Signer
+	PEM() []byte
+}
+
+type signerWithPEM struct {
+	crypto.Signer
+	pemData []byte
+}
+
+func (s signerWithPEM) PEM() []byte {
+	return s.pemData
+}
+
+// CA is a certificate authority usable to generated signed certificates
+type CA interface {
+	PrivateKey() SignerWithPEM
+	Cert() *x509.Certificate
+	NewSignedCert(cfg certutil.Config, publicKey crypto.PublicKey) (*x509.Certificate, error)
+}
+
+type authority struct {
+	pvk  SignerWithPEM
+	cert *x509.Certificate
+}
+
+func (ca *authority) PrivateKey() SignerWithPEM {
+	return ca.pvk
+}
+
+func (ca *authority) Cert() *x509.Certificate {
+	return ca.cert
+}
+
+func (ca *authority) NewSignedCert(cfg certutil.Config, publicKey crypto.PublicKey) (*x509.Certificate, error) {
+	serial, err := cryptorand.Int(cryptorand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	if len(cfg.CommonName) == 0 {
+		return nil, errors.New("must specify a CommonName")
+	}
+	if len(cfg.Usages) == 0 {
+		return nil, errors.New("must specify at least one ExtKeyUsage")
+	}
+	certTmpl := x509.Certificate{
+		Subject: pkix.Name{
+			CommonName:   cfg.CommonName,
+			Organization: cfg.Organization,
+		},
+		DNSNames:     cfg.AltNames.DNSNames,
+		IPAddresses:  cfg.AltNames.IPs,
+		SerialNumber: serial,
+		NotBefore:    ca.cert.NotBefore,
+		NotAfter:     time.Now().Add(time.Hour * 24 * 365).UTC(),
+		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:  cfg.Usages,
+	}
```

Comment:
I would add `BasicConstraintsValid: true` with `isCA: false` just to be sure to add the BasicConstraints extension